[docs] usecase: opentelemetry (#3439)
feldera/feldera: docs/use_cases/otel/representing_otel_data.md

# Representing OTel Data

> OpenTelemetry data is typically structured as nested JSON, which we can model into custom types in Feldera based on the [OTel Protobuf definitions](https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto). While it is possible to represent the entire OTel JSON data as a `VARIANT` type ([VARIANT docs](https://docs.feldera.com/sql/json#the-variant-type)), it is often more manageable to use custom types for easier querying.

Reviewer: custom data types are more compact and offer better typechecking

```sql

> CREATE TYPE KeyValue AS (

Reviewer: you have to explain that these definitions are SQL type definitions in Feldera SQL, which allows you to define custom user-defined types. Not all SQL dialects can do that.

    key VARCHAR,

>     value VARIANT

Reviewer: explain that a VARIANT Is still used because it can behave like a union type.

Author: The value is a JSON. Something like this: `{"stringValue": "some value"}`.

);

CREATE TYPE Event AS (

>     timeUnixNano VARCHAR,

Reviewer: char?

CREATE TYPE Metric AS (
    name VARCHAR,
    description VARCHAR,
    unit VARCHAR,
    data VARIANT,
    metadata KeyValue ARRAY
);

CREATE TYPE LogRecords AS (
    attributes KeyValue ARRAY,
    timeUnixNano VARCHAR,
    observedTimeUnixNano VARCHAR,
    severityNumber INT,
    severityText VARCHAR,
    flags INT4,
    traceId VARCHAR,
    spanId VARCHAR,
    eventName VARCHAR,
    body VARIANT
);

CREATE TYPE Scope AS (
    name VARCHAR,
    version VARCHAR,
    attributes KeyValue ARRAY
);

CREATE TYPE ScopeSpans AS (
    scope Scope,
    spans Span ARRAY
);

CREATE TYPE ScopeLogs AS (
    scope Scope,
    logRecords LogRecords ARRAY
);

CREATE TYPE ScopeMetrics AS (
    scope Scope,
    metrics Metric ARRAY
);

CREATE TYPE Resource AS (
    attributes KeyValue ARRAY
);

CREATE TYPE ResourceMetrics AS (
    resource Resource,
    scopeMetrics ScopeMetrics ARRAY
);

CREATE TYPE ResourceSpans AS (
    resource Resource,
    scopeSpans ScopeSpans ARRAY
);

CREATE TYPE ResourceLogs AS (
    resource Resource,
    scopeLogs ScopeLogs ARRAY
);
```

The following graph illustrates the type hierarchy of the custom types defined above:

![OTel Type Heirarchy](otel-type-heirarchy.png)

> Now that we have the type definitions to represent the OTel data, we create tables.

Reviewer: you have to explain which box you are programming again.
you also have to say that in Feldera tables correspond to input streams, they are not real "tables" like in other databases.

```sql
-- Input table that ingests resource spans from the collector.
CREATE TABLE otel_traces (
    resourceSpans ResourceSpans ARRAY
) WITH ('append_only' = 'true');

-- Input table that ingests resource logs from the collector.
CREATE TABLE otel_logs (
    resourceLogs ResourceLogs ARRAY
) WITH ('append_only' = 'true');

-- Input table that ingests resource metrics from the collector.
CREATE TABLE otel_metrics (
    resourceMetrics ResourceMetrics ARRAY
) WITH ('append_only' = 'true');
```

> Feldera operates on changes, so any input may be an insertion or deletion (Internally, Feldera represents these changes as [Z-sets](https://www.feldera.com/blog/database-computations-on-z-sets)). Setting `'append_only' = 'true'`, ensures only insertions are supported for this table.

Reviewer: Z-sets are not really necessary for this discussion, keep it as simple as possible.
You can say that Feldera can sometimes optimize programs better if it knows that no deletions will happen.
(BTW: I don't think we enforce the fact that there are no deletions, we trust the user.)

# Representing OTel Data

> ![Definition Custom Types to Model OTel data in Feldera](feldera-type-definitions.png)

Reviewer 2: ```suggestion
![Custom Type Definitions to Model OTel data in Feldera](feldera-type-definitions.png)
```

> OpenTelemetry data is typically structured as nested JSON, which we can model into custom types in Feldera

Reviewer 2: ```suggestion
OpenTelemetry data is typically structured as nested JSON, which we can model as used-defined SQK types in Feldera
```

Add a reference to the user-defined types section in SQL docs

based on the [OTel Protobuf definitions](https://github.com/open-telemetry/opentelemetry-proto/tree/main/opentelemetry/proto).
While it is possible to represent the entire OTel JSON data as a `VARIANT` type ([VARIANT docs](https://docs.feldera.com/sql/json#the-variant-type)),

> it is often more manageable to use custom types for easier querying.

Reviewer 2: ```suggestion
user-defined types are more efficient and ergonomic in cases when the JSON schema is known in advance.
```

> Custom types are also more compact, and offer better typechecking.

Reviewer 2: ```suggestion
Custom types also have a smaller memory footprint and offer better type checking.
```

    metadata KeyValue ARRAY
);

CREATE TYPE LogRecords AS (
    attributes KeyValue ARRAY,
    timeUnixNano CHAR(20),
    observedTimeUnixNano CHAR(20),
    severityNumber INT,
    severityText VARCHAR,
    flags INT4,
    traceId VARCHAR,
    spanId VARCHAR,
    eventName VARCHAR,
    body VARIANT
);

CREATE TYPE Scope AS (
    name VARCHAR,
    version VARCHAR,
    attributes KeyValue ARRAY
);

CREATE TYPE ScopeSpans AS (
    scope Scope,
    spans Span ARRAY
);

CREATE TYPE ScopeLogs AS (
    scope Scope,
    logRecords LogRecords ARRAY
);

CREATE TYPE ScopeMetrics AS (
    scope Scope,
    metrics Metric ARRAY
);

CREATE TYPE Resource AS (
    attributes KeyValue ARRAY
);

CREATE TYPE ResourceMetrics AS (
    resource Resource,
    scopeMetrics ScopeMetrics ARRAY
);

CREATE TYPE ResourceSpans AS (
    resource Resource,
    scopeSpans ScopeSpans ARRAY
);

CREATE TYPE ResourceLogs AS (
    resource Resource,
    scopeLogs ScopeLogs ARRAY
);
```

The following graph illustrates the type hierarchy of the custom types defined above:

![OTel Type Hierarchy](otel-type-hierarchy.png)

Now that we have the type definitions to represent the OTel data, we create tables.

> Tables in Feldera correspond to input streams and aren't **real tables** like in other databases.

Reviewer 2: ```suggestion
Tables in Feldera model input data streams.
```

I know this was @mihaibudiu 's suggestion but most people won't know what we mean by saying they are not real tables. We can insert into and query these tables, which makes them pretty real actually.
